### PR TITLE
Implement Identity services and token generation

### PR DESCRIPTION
**Pull request by MahdiyarGHD**

### CODE REVIEW

File: src/CSharp/EasyMicroservices.IdentityMicroservice.Domain/Contracts/Responses/UserResponseContract.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using EasyMicroservices.Cores.Interfaces;
+
+namespace EasyMicroservices.IdentityMicroservice.Contracts.Responses
+{
+    public class UserResponseContract : IUniqueIdentitySchema
+    {
+        public string Token { get; set; }
+        public string UniqueIdentity { get; set; }
```
Comment (Ali-YousefiTelori) [inline src/CSharp/EasyMicroservices.IdentityMicroservice.Domain/Contracts/Responses/UserResponseContract.cs]: We do not use UID in user response

File: src/CSharp/EasyMicroservices.IdentityMicroservice.Domain/EasyMicroservices.IdentityMicroservice.Domain.csproj [hunk 2]
```diff
@@ -1,20 +1,23 @@
...
 
   <ItemGroup>
-    <PackageReference Include="EasyMicroservices.Cores.AspEntityFrameworkCoreApi" Version="0.0.0.45" />
-    <PackageReference Include="EasyMicroservices.Cores.Database" Version="0.0.0.45" />
+	  <PackageReference Include="EasyMicroservices.Cores.AspEntityFrameworkCoreApi" Version="0.0.0.49" />
+	  <PackageReference Include="EasyMicroservices.Cores.Database" Version="0.0.0.49" />
   </ItemGroup>
 
+	<ItemGroup>
+		<Reference Include="EasyMicroservices.AuthenticationsMicroservice.Clients">
```
Comment (Ali-YousefiTelori) [inline src/CSharp/EasyMicroservices.IdentityMicroservice.Domain/EasyMicroservices.IdentityMicroservice.Domain.csproj]: We have to remove this direct use

File: src/CSharp/EasyMicroservices.IdentityMicroservice.Logics/Helpers/IdentityHelper.cs [hunk 3]
```diff
@@ -28,20 +30,51 @@ public IdentityHelper(IConfiguration config, IJWTManager jwtManager)
...
         }
 
-        //public virtual async Task<MessageContract<long>> Register(AddUserRequestContract input)
-        //{
-        //    string Password = input.Password;
-        //    input.Password = await SecurityHelper.HashPassword(input.Password);
 
-        //    //var logic = _unitOfWork.GetLongContractLogic<UserEntity, AddUserRequestContract, UserContract, UserContract>();
-        //    //var usersRecords = await logic.GetBy(x => x.UserName == input.UserName.ToLower());
+        public async Task<MessageContract<long>> Register(Contracts.Requests.AddUserRequestContract request)
```
Comment (Ali-YousefiTelori) [inline src/CSharp/EasyMicroservices.IdentityMicroservice.Logics/Helpers/IdentityHelper.cs]: Please create RegisterResponseContract for this method

File: src/CSharp/EasyMicroservices.IdentityMicroservice.Logics/Helpers/IdentityHelper.cs [hunk 4]
```diff
@@ -28,20 +30,51 @@ public IdentityHelper(IConfiguration config, IJWTManager jwtManager)
...
+            var user = await _userClient.AddAsync(new AddUserRequestContract
+            {
+                UserName = request.UserName,
+                Password = request.Password
+            });
+
+            return user.Result;
+        }
+
+        public virtual async Task<MessageContract<long>> Login(Contracts.Common.UserSummaryContract cred)
```
Comment (Ali-YousefiTelori) [inline src/CSharp/EasyMicroservices.IdentityMicroservice.Logics/Helpers/IdentityHelper.cs]: Please create LoginResponseContract for this method

File: src/CSharp/EasyMicroservices.IdentityMicroservice.Logics/Helpers/IdentityHelper.cs [hunk 5]
```diff
@@ -28,20 +30,51 @@ public IdentityHelper(IConfiguration config, IJWTManager jwtManager)
...
+        {
+            var user = await _userClient.VerifyUserIdentityAsync(new Authentications.GeneratedServices.UserSummaryContract { UserName = cred.UserName, Password = cred.Password});
+            if (!user.IsSuccess)
+                return (ServiceContracts.FailedReasonType.AccessDenied, "Username or password is invalid."); //"Username or password is invalid."
```
Comment (Ali-YousefiTelori) [inline src/CSharp/EasyMicroservices.IdentityMicroservice.Logics/Helpers/IdentityHelper.cs]: We have Incorrect FailedReason, don't use AccessDenied.

**Review by Hamed-Bagherpour — APPROVED** [review]

File: src/CSharp/EasyMicroservices.IdentityMicroservice.WebApi/Controllers/IdentityController.cs [hunk 6]
```diff
@@ -1,23 +1,123 @@
...
+        {
+            _appUnitOfWork = appUnitOfWork;
+            _config = _appUnitOfWork.GetConfiguration();
+            _authRoot = _config.GetValue<string>("RootAddresses:Authentications");
+            _userClient = new(_authRoot, new System.Net.Http.HttpClient());
+            _identityHelper = _appUnitOfWork.GetIdentityHelper();
+        }
+
+        [HttpPost]
+        public async Task<MessageContract<bool>> VerifyUserName(VerifyEmailAddressContract request)
```
Comment (Ali-YousefiTelori) [inline src/CSharp/EasyMicroservices.IdentityMicroservice.WebApi/Controllers/IdentityController.cs]: Please add VerifyUserNameResponseContract.
Why VerifyUserName has VerifyEmailAddressContract for request?

File: src/CSharp/EasyMicroservices.IdentityMicroservice.WebApi/Controllers/IdentityController.cs [hunk 7]
```diff
@@ -1,23 +1,123 @@
...
+                Password = user.Result.Password,
+                UniqueIdentity = user.Result.UniqueIdentity,
+                UserName = user.Result.UserName,
+            }).AsCheckedResult();
+
+            return true;
+        }
+
+        [HttpPost]
+        public async Task<MessageContract<long>> Register(Contracts.Requests.AddUserRequestContract request)
```
Comment (Ali-YousefiTelori) [inline src/CSharp/EasyMicroservices.IdentityMicroservice.WebApi/Controllers/IdentityController.cs]: Please add RegisterResponseContract

File: src/CSharp/EasyMicroservices.IdentityMicroservice.WebApi/Controllers/IdentityController.cs [hunk 8]
```diff
@@ -1,23 +1,123 @@
...
+        {
+            return await _identityHelper.Register(request);
+        }
 
-        public IdentityController()
+        [HttpPost]
+        public async Task<MessageContract<long>> Login(Contracts.Common.UserSummaryContract request)
```
Comment (Ali-YousefiTelori) [inline src/CSharp/EasyMicroservices.IdentityMicroservice.WebApi/Controllers/IdentityController.cs]: Please Add LoginResponseContract

**Review by Ali-YousefiTelori — CHANGES_REQUESTED** [review]

**Review by mahdidelzendeh72 — APPROVED** [review]
thanks

**Review by Ali-YousefiTelori — APPROVED** [review]

**Pull request merged**